subsys/serialization/transport/trans_rpmsg.c

```diff
@@ -0,0 +1,201 @@
+
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <errno.h>
+
+#include <rp_errors.h>
+
+#include "rp_ll.h"
+#include "trans_rpmsg.h"
+
+#include <logging/log.h>
+
+
+LOG_MODULE_REGISTER(trans_rpmsg);
+
+
+struct fifo_item {
+	void *fifo_reserved;
+	size_t size;
+};
+
+
+static rp_trans_receive_handler receive_handler;
+
+
+static int translate_error(int rpmsg_err)
+{
+	switch (rpmsg_err) {
+	case RPMSG_ERR_NO_MEM:
+		return RP_ERROR_NO_MEM;
+	case RPMSG_ERR_NO_BUFF:
+		return RP_ERROR_NO_MEM;
+	case RPMSG_ERR_PARAM:
+		return RP_ERROR_INVALID_PARAM;
+	case RPMSG_ERR_DEV_STATE:
+		return RP_ERROR_INVALID_STATE;
+	case RPMSG_ERR_BUFF_SIZE:
+		return RP_ERROR_NO_MEM;
+	case RPMSG_ERR_INIT:
+		return RP_ERROR_INTERNAL;
+	case RPMSG_ERR_ADDR:
+		return RP_ERROR_INTERNAL;
+	default:
+		if (rpmsg_err < 0) {
+			return RP_ERROR_INTERNAL;
+		}
+		break;
+	}
+	return RP_SUCCESS;
+}
+
+
+static void *stack_allocate(k_thread_stack_t **stack, size_t stack_size)
+{
+	u8_t *buffer =
+		k_malloc(K_THREAD_STACK_LEN(stack_size) + STACK_ALIGN + 1);
+	uintptr_t ptr = (uintptr_t)buffer;
+	uintptr_t unaligned = ptr % STACK_ALIGN;
+
+	if (buffer == NULL) {
+		*stack = NULL;
+		LOG_ERR("Endpoint thread stack allocation failed!");
+		__ASSERT(buffer, "Out of memory");
+	} else if (unaligned == 0) {
+		*stack = (k_thread_stack_t *)buffer;
+	} else {
+		*stack = (k_thread_stack_t *)(&buffer[STACK_ALIGN - unaligned]);
+	}
+	return buffer;
+}
+
+int rp_trans_init(rp_trans_receive_handler callback)
+{
+	int result;
+
+	receive_handler = callback;
+	result = rp_ll_init();
+	return translate_error(result);
+}
+
+void rp_trans_uninit(void)
+{
+	rp_ll_uninit();
+}
+
+static void endpoint_thread(void *p1, void *p2, void *p3)
+{
+	struct rp_trans_endpoint *endpoint = (struct rp_trans_endpoint *)p1;
+	struct fifo_item *item;
+
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	while (true) {
+		k_sem_take(&endpoint->sem, K_FOREVER);
+		if (!endpoint->running) {
+			break;
+		}
+		do {
+			item = k_fifo_get(&endpoint->fifo, K_NO_WAIT);
+			if (item == NULL) {
+				break;
+			}
+			receive_handler(endpoint, (const u8_t *)(&item[1]),
+				item->size);
+			k_free(item);
+		} while (true);
+	}
+}
+
+static void event_handler(struct rp_ll_endpoint *ep,
+	enum rp_ll_event_type event, const u8_t *buf, size_t length)
+{
+	struct rp_trans_endpoint *endpoint =
+		CONTAINER_OF(ep, struct rp_trans_endpoint, ep);
+
+	if (event == RP_LL_EVENT_DATA) {
+		struct fifo_item *item =
+			k_malloc(sizeof(struct fifo_item) + length);
+		if (!item) {
+			LOG_ERR("Out of memory when receiving incoming packet");
+			__ASSERT(item, "Out of memory");
+		}
+		memcpy(&item[1], buf, length);
+		item->size = length;
+		k_fifo_put(&endpoint->fifo, item);
+	}
```

Comment:
There is at least one rx thread too many in this module and rp_ll.c.

The memory we're receiving in this callback is already safely allocated in the ring buffer in the shared memory. Unless there are some restrictions in the shared memory that forces us to do this, I think we should call user `receive_handler` directly here, while the memory is still guaranteed to be allocated. 

I can see that this is done to be able to run each end point in a different stack with a user defined stack size, but I think this approach creates more problems than it solves. By choosing to own the user threads in this module, we're forced to allocate the thread's stack on the heap on line 141, and generally create a lot of complexity and unnecessary points of failure.

I suggest moving the thread definition into the header of this file:

```C
struct rp_trans_endpoint {
	struct k_thread thread;
	k_thread_stack_t *stack;
	struct k_sem thread_sem;
	struct k_sem data_sem;
	struct {
		u8_t *buf;
		size_t len;
	} evt;
	rp_trans_receive_handler cb;
};

#define RP_TRANS_ENDPOINT(_name, _stack_size, _handler)                        \
	K_THREAD_STACK_DEFINE(_rp_trans_stack_##_name, _stack_size);           \
	static struct rp_trans_endpoint _name = {                              \
		.cb = _handler,                                                \
		.stack = _rp_trans_stack_##_name,                              \
	};

```

This removes the need for allocating the stack on the heap.

Then, inside this file, define the `rp_trans_thread` thread function:

```C
static void rp_trans_thread(void *p1, void *p2, void *p3)
{
	struct rp_trans_endpoint *ept = p1;

	while (1) {
		k_sem_take(&ept->thread_sem, K_FOREVER); // or something less banal
		ept->cb(ept, ept->evt.buf, ept->evt.len);
		k_sem_give(&ept->data_sem, K_FOREVER);
	}
}
```

This just calls the callback really, making sure we have data in the event buffer.

Then, in the event handler, instead of copying the buffer into a fifo, just pass it to the user thread without buffering, and block until it's done processing:

```C
static void event_handler(struct rp_ll_endpoint *ep,
	enum rp_ll_event_type event, const u8_t *buf, size_t length)
{
	struct rp_trans_endpoint *endpoint =
		CONTAINER_OF(ep, struct rp_trans_endpoint, ep);

	if (event == RP_LL_EVENT_DATA) {
		endpoint->evt.buf = buf;
		endpoint->evt.len = len;
		k_sem_give(&ept->thread_sem, K_FOREVER);
		k_sem_take(&ept->data_sem, K_FOREVER);
	}
}
```

This could probably be refined a bit, and it'll force us to process the events for each endpoint in order, but with the current user code, this isn't really a problem. It definitely beats randomly asserting 2 months into the deployment because we failed to allocate from heap.

Reply:
> just pass it to the user thread without buffering, and **block until it's done processing**

This is the main reason why we need multiple threads and fifos. We cannot block until it's done processing. Each endpoint may handle different API, e.g. BLE on endpoint 0, user defined API on endpoint 1. If user calls some long running function and wait it for e.g. 10 sec. then BLE API is completely disable during this period. This may result e.g. unexpected disconnection.

> I suggest moving the thread definition into the header of this file:
> This removes the need for allocating the stack on the heap.

Yes, this solution looks good. I will consider adding it.

Reply:
Okay. I wish we could enforce non-blocking usage, but since we're trying to wrap upstream and user APIs, I suppose it's going to be hard to make such a restriction. 

I still **really** want to get rid of this malloc though (or the assert, really). We can't rely on the heap for such a critical module, IMO, it's virtually guaranteed to run out of memory at some point for some deployments, likely several hours/days/weeks after bootup.

Second suggestion, then: 

Currently, each endpoint only allows a single command at the time, as it only holds one response handler, and is waiting for the semaphore before moving on. This seems like a reasonable limitation to me, so lets take advantage of it. 

This means that we'll only need one command buffer per endpoint at the handler side. Allocate this buffer statically in the endpoint structure (with user defined size, like the stack, or put the it on stack in the handler thread), and copy into it from the worker thread. The semaphore on the emitter side creates a flow control mechanism, that prevents us from sending commands we don't have memory for on the handler side. Emitter side flow control is the only safe way to deal with this, as it's too late to do something about the lack of memory in the handler thread (as demonstrated by the assert in this function).

Currently, we're able to emit events while pending on a command in another thread. In single-core applications, events are normally blocking operations (ie callbacks), and IMO, we can do the same here, blocking events on a per-thread level. If we block on the rp_ser endpoint semaphore for the events too, we'll be able to move the blockage up into each endpoint event emitter thread. This adds flow control for the events as well.

Command responses can either be allocated in a static buffer in each end point, or (even better) in a stack-allocated buffer in the blocked command sender thread.

We can also set up a static queue for each endpoint if we really need the events to go out asynchronously, but if we do, we should still do the blocking in the emitter thread when memory runs out. This would require us to communicate the allocated queue size in the handshake.

My key point is that we should establish per-endpoint flow control on the emitter side, instead of making the handler side worker thread block (my first suggestion) or crash (like we do now). If we prevent pushing to the shared linked list while the endpoint is blocked, we avoid blocking other end points.

Reply:
Some time ago I created a solution (in the paper so far) that looks similar to your description, but finally we go with a solution that you see right now in the code. I will discuss with my colleagues in my team if we want to change our approach.

Reply:
Ok, so I presented to colleagues a different approach that I mentioned in a previous comment. We agreed that it is something that we can implement in this PR.
This change will remove need for fifo, malloc and memory copying. It takes advantage of waiting for response (something that you already noticed).

### So how it will work:

Lets assume that we have following threads:
* *[user thread]* - some thread from which user calls our API functions that will be serialized.
* *[endpoint thread]* - thread that is assigned to handle specific endpoint (instance of rp_ser).
* *[rx thread]* - common thread that receives all incoming packets from all endpoints. We want to block it as little as possible.

Calling a remote function:
1. The input parameters are encoded into a buffer. Buffer is allocated by the transport layer and it is up to the transport how it is allocated, e.g. for rpmsg it is allocated on the caller stack *[user thread]*
1. Information how to interpret response is set. Information contains two pointers: a pointer to function responsible for decoding (decoder) and a pointer where to put the results of the decoding. *[user thread]*
1. Buffer is send to the other core as a `command` packet *[user thread]*
1. Waiting for response *[user thread]*
1. Function is executed and response is generated *[other core]*
1. `response` packet is received by rx thread *[rx thread]*
1. Decoder is executed, so packet is decoded in context of rx thread, so no additional copying or queuing is required. Bloking rx thread just for time of decoding is still ok. *[rx thread]*
1. User thread is notified that response was decoded *[rx thread]*
1. Return from a function *[user thread]*

The other core called a function in this core:
1. Rx thread gets a `command` or an `event` packet *[rx thread]*
1. Rx thread wakes up the endpoint thread and provides a packet as a pointer to avoid copying. *[rx thread]*
1. Rx thread is waiting. This is a potential place where rx thread may wait for too long, so the other core is responsible to taking care that `command` or `event` is not send until previous one is completed. *[rx thread]*
1. Endpoint thread decodes the packet *[endpoint thread]*
1. Endpoint thread unblocks the rx thread *[endpoint thread]*
1. Rx thread returns packet handling function *[rx thread]*
1. Target function is executing *[endpoint thread]*
1. If this was `command` then response is encoded into a buffer. Buffer is allocated by the transport layer and it is up to the transport how it is allocated, e.g. for rpmsg it is allocated on the caller stack *[endpoint thread]*
1. `response` or `event_ack` is send to the other core *[endpoint thread]*

I created some [diagrams on draw.io](https://www.draw.io/#Hdoki-nordic%2Fdraw.io%2Fdrafts%2FNo%20memory%20copy%20concept.drawio) to visualize it.

Reply:
This sounds like a big improvement on the current implementation.

> 3. Rx thread is waiting. This is a potential place where rx thread may wait for too long, so the other core is responsible to taking care that command or event is not send until previous one is completed. [rx thread]
> 4. Endpoint thread decodes the packet [endpoint thread]
> 5. Endpoint thread unblocks the rx thread [endpoint thread]

This only works if the decoders are non-blocking, ie they decode the packet fully before yielding the buffer, *then* call whatever API they represent. This needs to be properly communicated in the decoder implementation guide and highlighted by the sample. Currently, the entropy_ser sample does not follow this requirement. 

> `event_ack`

With this construct, I suppose there's no functional difference between commands and events for the transport layer anymore. An event_ack is just a command response without data, no? Could the decoders be merged in rp_ser, and the `rp_ser_packet_type` reduced to cmd and rsp?

Reply:
Yes, decoders be merged, and no, `rp_ser_packet_type` cannot be reduced to cmd and rsp.

The main different between `command` and `event` is that thread calling `command` waits for the `response` and thread sending `event` can exit immediately after sending packet. `event_act` construct is to inform rp_ser that the other side is ready to receive next `command` or `event`.